### PR TITLE
Add items, auth_services, end_services, and auth_ip_addresses 

### LINKED ISSUE #2: Add API Authentication

### PR DESCRIPTION
**Pull request by grepsedawk**
- Adds all models/controllers/routes for an authentication API
- Tosses in a few extras, like `better_errors` and `guard-rspec`
- Implements testing for Ruby `2.3.3` as well

resolves #2

### CODE REVIEW

**Comment by ekate** [comment]
@pachok here are my initial comments:
1.I saw you've open an issue to add mysql instead of sqlite. I would  expect you to do the change in that pull request.
2. In the database schema you've omitted the code column for auth_service and end_service. De-normalization might be a controversial issue so I didn't insist on it but I still want to see human generated 4 letter codes in the lookup tables. They will be used as query parameters. Also in ip_addresses table you decided to use de-normalization, e.g. column "name". In that particular case lookup table will make more sense. ip addresses are associated with different rooms and it is useful to be able to query by room and this query doesn't need to be fast
3. Why do you have api/v1 directory instead of putting everything api. That is internal application so we are not going to support versioning
4.As a minimum in the item controller "all" method should support pagination. Ideally it should be done for all controllers that list all records
5. I am not sure we need to use guard as usually it's not the part of our workflow, but I will let the whole group to decide what they think

**Comment by ekate** [comment]
@pachok, I see you've just added mysql while I was writing a comment. Thank you

**Comment by grepsedawk** [comment]
1. Sorry, I rebased master while you were working on the comment.
2.
    a. No problem, I can add the code column in a later migration. I didn't know how much we wanted to keep these human generated with it being a purely API based app.
    b. As far as the ip address, I actually switched the relationship to a more natural relationship. `auth_service` now ONLY relates to `auth_service`s (Shib, Facebook, Google, etc) and `auth_ip_address`es (172.34.235.21, 10.30.53.11, etc) relate directly to items.
3. I used `api/v1` because, if NYU is going to potentially have many different apps using this application, it makes sense to have the version namespace on the API. Even if we think we'll never need it, it's easier to keep the `v1` in the case NYU later wants to later support an older version of this API. The way I see it is it doesn't hurt to keep it there, but could hurt later. That being said, if we still want the `/v1` removed, let me know and I will remove it.
4. Added.
5. I use guard exclusively for rspec. If this is something we don't want, it can be removed pretty easily at any time without issue.

**Comment by ekate** [comment]
2 yes, ips can be related directly to items but ip can be grouped by rooms, so I am asking you to add this lookup table
3 Please remove versioning

**Comment by grepsedawk** [comment]
2. I just want to ensure I'm clear on what you're asking to do. Are you asking to do the comparable thing to IP's that I was asking (about auth_services) here: https://nyucomposers.slack.com/archives/general/p1486650386000021 ?
3. Wilco

File: app/controllers/api/api_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  private
+  def authenticate!
+    unless params.has_key?(:api_key) && params[:api_key].length == 48
+      head :unauthorized, status: 401
+      return false
+    end
+
+    api_key_public = params[:api_key][0..23]
+    api_key_private = params[:api_key][24..48]
```
Comment (jgpawletko) [inline app/controllers/api/api_controller.rb]: It looks like your total `api_key` length is 48 characters. Therefore, due to zero indexing, I believe this should be `params[:api_key][24..47]`.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -10,4 +10,10 @@ def authentication_callback
   def authenticate!
     redirect_to '/auth/nyulibraries' unless session[:auth] && (session[:auth]['credentials']['expires'] == false || session[:auth]['credentials']['expires_at'] > Time.now.to_i)
```
Comment (jgpawletko) [inline app/controllers/application_controller.rb:11]: Hi @pachonk ,
You might want to create a private helper method to encapsulate the logic after `unless`.
i.e., 
```
def valid_session?
  session[:auth] && 
    (session[:auth]['credentials']['expires'] == false || 
      session[:auth]['credentials']['expires_at'] > Time.now.to_i)
end
```
then your `authenticate!` method could be:
```
...
  redirect_to '/auth/nyulibraries' unless valid_session?
...
```

File: db/migrate/20170217185047_create_api_keys.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class CreateApiKeys < ActiveRecord::Migration[5.0]
+  def change
+    create_table :api_keys do |t|
+      t.string :api_key_public
+      t.string :api_key_hash
```
Comment (jgpawletko) [inline db/migrate/20170217185047_create_api_keys.rb:5]: is `:api_key_hash` the same as the private fragment of `params[:api_key]` or is there some transformation between the database attribute and the value used in the controller?

File: lib/ipv4_address_serializer.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class IPv4AddressSerializer
+  def self.load(value)
+    IPAddr.new(value, Socket::AF_INET).to_s if value
+  end
+
+  def self.dump(value)
+    IPAddr.new(value).to_i
+  end
+end
```
Comment (jgpawletko) [inline lib/ipv4_address_serializer.rb:9]: I know this is a picky comment, but it looks like you're missing a newline

File: lib/ipv6_address_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class IPv6AddressSerializer
+  def self.load(value)
+    IPAddr.new_ntoh(value).to_s if value
+  end
+
+  def self.dump(value)
+    IPAddr.new(value).hton
+  end
+end
```
Comment (jgpawletko) [inline lib/ipv6_address_serializer.rb:9]: I know this is a picky comment, but it looks like you're missing a newline

File: spec/controllers/api/api_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    context 'invalid API key' do
+      it do
+        get :index, params: { api_key: 'invalid_api_key' }
+        expect(response).to have_http_status :unauthorized
+      end
+    end
+  end
+end
```
Comment (jgpawletko) [inline spec/controllers/api/api_controller_spec.rb:33]: I know this is a picky comment, but it looks like you're missing a newline

**Review by jgpawletko — COMMENTED** [review]
Hi Alex,
I'm still reviewing your code, but thought I'd pass along the comments I have so far.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (grepsedawk) [reply app/controllers/application_controller.rb:11]: Yeah, I had definitely considered that, but the massive `unless` was temporary until we decided what would be considered "authenticated" to authz. In the next PR I intend to move that to a new function.

File: db/migrate/20170217185047_create_api_keys.rb (hunk 3 above)
Comment (grepsedawk) [reply db/migrate/20170217185047_create_api_keys.rb:5]: `api_key_hash` is (pseudo code) `hash(api_key_public + api_key_secret)`

File: lib/ipv6_address_serializer.rb (hunk 5 above)
Comment (grepsedawk) [reply lib/ipv6_address_serializer.rb:9]: Not picky at all. Will fix!

File: lib/ipv4_address_serializer.rb (hunk 4 above)
Comment (grepsedawk) [reply lib/ipv4_address_serializer.rb:9]: Good catch, fixed.

File: app/controllers/api/api_controller.rb (hunk 1 above)
Comment (grepsedawk) [reply app/controllers/api/api_controller.rb]: Fixed, thanks!

**Comment by jgpawletko** [comment]
Hi @pachonk ,
I don't want to hold this up anymore, but I do want to look at an ERD to make sure I understand how the schema hangs together.  

Also, do you see advantages to using `has_and_belongs_to_many` vs.  with `has_many :through` ?  I usually prefer `has_many :through` because it allows one to add additional attributes to associations if needed.

Thanks!

Hi @ekate, 
feel free to merge if you're OK with the PR.

Thanks!

**Pull request merged**